src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -382,6 +382,11 @@ public void OnStarting(Func<object, Task> callback, object state)
         {
             lock (_onStartingSync)
             {
+                if (HasResponseStarted)
+                {
+                    ThrowResponseAlreadyStartedException(nameof(OnStarting));
```

Comment:
@benaadams Could you elaborate on your first comment about using an enum? Not sure what's going on there.